packages/cli/src/modules/chat-hub/chat-hub.service.ts

```diff
@@ -297,56 +337,254 @@ export class ChatHubService {
+			runIndex,
+		);
+
+		const workflow = await this.createChatWorkflow(
+			user,
+			session.id,
+			history,
+			message,
+			payload.credentials,
+			payload.model,
+		);
+
+		await this.executeChatWorkflow(
+			res,
+			user,
+			workflow,
+			replyId,
+			sessionId,
+			messageId,
 			turnId,
-			previousMessageId: payload.previousMessageId ?? null,
-			...usedModel,
+			selectedModel,
+		);
+	}
+
+	async regenerateAiMessage(res: Response, user: User, payload: RegenerateMessagePayload) {
+		const { sessionId, retryId, replyId } = payload;
+
+		const selectedModel: ModelWithCredentials = {
+			...payload.model,
+			credentialId: this.getCredentialId(payload.model.provider, payload.credentials),
+		};
+
+		const session = await this.getChatSession(user, payload, selectedModel, false);
+		const messages = session.messages ?? [];
+		const messageToRetry = await this.getChatMessage(session.id, retryId);
+
+		if (messageToRetry.type !== 'ai') {
+			throw new BadRequestError('Can only retry AI messages');
+		}
+
+		const historyIds = messageToRetry.previousMessageId
+			? this.buildActiveMessageChain(messages, messageToRetry.previousMessageId)
+			: [];
+
+		const history = historyIds.flatMap((id) => {
+			return messages.find((m) => m.id === id) ?? [];
 		});
 
+		const lastHumanMessage = history.filter((m) => m.type === 'human').pop();
+		if (!lastHumanMessage) {
+			throw new BadRequestError('No human message found to base the retry on');
+		}
+
+		// Remove any (AI) messages that came after the last human message
+		const lastHumanMessageIndex = history.indexOf(lastHumanMessage);
+		if (lastHumanMessageIndex !== -1) {
+			history.splice(lastHumanMessageIndex + 1);
+		}
+
+		// Rerun the workflow, replaying the last human message
+		const workflow = await this.createChatWorkflow(
+			user,
+			session.id,
+			history,
+			lastHumanMessage ? lastHumanMessage.content : '',
+			payload.credentials,
+			payload.model,
+		);
+
+		// If the message being retried is itself a retry, we want to point to the original message
+		const retryOfMessageId = messageToRetry.retryOfMessageId ?? messageToRetry.id;
+		const otherRuns = messages.filter((m) => m.retryOfMessageId === retryOfMessageId);
+		const runIndex = otherRuns.length + 1;
+
+		await this.messageRepository.updateChatMessage(retryOfMessageId, { state: 'replaced' });
+		for (const run of otherRuns) {
+			if (run.state === 'active') {
+				await this.messageRepository.updateChatMessage(run.id, { state: 'replaced' });
+			}
+		}
+
+		await this.executeChatWorkflow(
+			res,
+			user,
+			workflow,
+			replyId,
+			sessionId,
+			lastHumanMessage.id,
+			messageToRetry.turnId,
+			selectedModel,
+			retryOfMessageId,
+			runIndex,
+		);
+	}
+
+	private async executeChatWorkflow(
+		res: Response,
+		user: User,
+		workflow: {
+			workflowData: IWorkflowBase;
+			startNodes: StartNodeData[];
+			triggerToStartFrom: { name: string; data?: ITaskData };
+		},
+		replyId: ChatMessageId,
```

Comment:
I feel positional arguments are error prone when there are many args that share the same type like this

Reply:
yeah, I'll probably refactor these into some objects that we pass around